### DISCUSSION TITLE
Portalverbund - Berechtigungsänderung der Rolle eines Nutzers greift nicht - ist cache/clear eine Lösung?

### DISCUSSION THREAD

**Discussion by faulgurke** [e-netze/webgis-community#182] [Q&A]
Im WebGISWorkshop vom 03.07.2024 habe ich in die Runde gefragt, ob jemand folgendes Problem kennt: Ein Sachbearbeiter eines Verwaltungsverbundes hat mehrere Rollen für die Bearbeitung von Flächenwidmungen unterschiedlicher Gemeinden.
Beim Wechsel der Rolle greift die Berechtigung nicht sofort, sondern erst nach einer längeren Wartezeit.
Ich habe eure Vorschläge (Rolle wechseln, im trace.log verfolgen was ankommt, ...) berücksichtigt. 
Das Löschen des BrowserCaches und Neuanmeldung haben keine Wirkung.

Was hilft ist ein api/cache/clear !

Meine Frage ist jetzt, ob das Ausführen eines "api/cache/clear" eines WebGIS Nutzers (also eines Sachbearbeiters eines Verwaltungsverbundes) die Verwendung von WebGIS durch andere Nutzer beeinträchtigt.
Gibt es Probleme für andere Nutzer, die die Anwendung gleichzeitig verwenden wenn einer ein "api/cache/clear" ausführt?

**Comment by gis-eni** [comment]
Ein Cache/clear kannst du grundsätzlich immer machen. Dabei wird in Hintergrund das CMS neu eingelesen und wenn alles fertig ist, wir der "cache" ausgetauscht. Zu einer Beeinträchtigung kommt es nicht.

Kommt mir trotzdem seltsam vor, dass das sie Lösung ist.

Frage: in der api.config und in der portal.config hast du jeweils die Einträge:

``` xml
    <add key="cache-provider" value="fs" />
    <add key="cache-connectionstring" value="C:\temp\webgis5_db\cache" />
```
Hast du da in beiden Configs das gleiche stehen? Und wenn ja, welchen ``cache-provider`` hat ihr da eingestellt?
Hast du vielleicht auch noch entwas in der Art konfiguriert?

``` xml
    <add key="cache-aside-provider" value="redis" />
    <add key="cache-aside-connectionstring" value="localhost:6379" />
```

**Reply by faulgurke** [reply]
Für cache-provider und cache-connectionstring sieht es genauso aus wie von dir beschrieben.
In beiden Configs steht dasselbe, und der cache-provider ist "fs".

In der "api.config" habe ich einen cache-aside-provider definiert, in der portal.config **nicht**.

    <add key="cache-aside-provider" value="inapp" />    
    <add key="cache-aside-connectionstring" value="3600" />

**Reply by gis-eni** [reply]
Ok, das ist des Rätsels Lösung.
WebGIS Portal und API teilen sich einen Cache, um User Credentials serverseitig auszutauschen.

Die Konstellation, die du da hast ist gut für den Atlas, da steigt in der Regen eh nur der "anonyme" Benutzer ein, der keine Rollen hat.
Die API muss nicht bei jedem Request auf den Cache zugreifen, sondern merkt sich den "anonym" User im RAM.

In einer "Intranet" Umgebung, wo alle User Rollen haben, funktioniert der Ansatz auch allerdings merkt sich hier die API die Rollen für den User im RAM. Bekommt ein User neue Rollen, werden die zwar im Cache (bei dir Filesystem) neu gesetzt. Die API verwendet aber immer noch den Aside-Cache (Performance!). Der ist bei dir so eingestellt, dass Einträge 1 Stunde drinnen bleiben. Erst dann holt sich die API Werte wieder aus dem eigentlichen Cache.

In solchen Fällen müssen sich die Anwendungen auch den Aside Cache teilen! Oder du lässt den Aside Cache weg! Dann greift die API halt bei jedem Request auf den Cache (bei dir Filesystem) zu.

Empfehlung für Atlas ohne berechtigte User: So wie du das hast

Empfehlung für Intranet (berechtigte User) mit Rollen die sich ändern können:
* Cache sollte in einer Datenbank oder im Filesystem liegen
* Bei Filesystem (zugriffe sind in der Regel oft langsamer als auf eine Datenbank: darum hier einen schnellen Aside Cache verwenden. Diese muss aber für Portal und API der gleiche sein. Gut wäre hier etwa Redis. 

Fazit: Bring deinen Cache für diese Anwendung erst einmal in eine Datenbank und lass den AsideCache weg.

**Reply by faulgurke** [reply]
Zunächst vielen Dank für die ausführliche Erläuterung! Da habe ich schon wieder einiges dazugelernt.
Wenn ich deine Empfehlungen umsetzen möchte, heißt das für mich:
Intranet-Umgebung (Portal mit berechtigten usern)
1. Umstellung des "cache-providers" auf value="db" und Einrichtung einer DB-Tabelle für das caching
2. Deaktivierung des "cache-aside-provider" in der api.config (in der portal.config gibt es ihn eh nicht).

Richtig?

**Reply by gis-eni** [reply]
Genau so!
